### ISSUE TITLE
PODMAN_ALLOW_SINGLE_ID_MAPPING_IN_USERNS=1 causes podman to hang

### ISSUE THREAD

**Issue by llchan** [containers/libpod#1625]
**kind bug**

**Description**

I am running rootless podman with 
`PODMAN_ALLOW_SINGLE_ID_MAPPING_IN_USERNS=1` and no /etc/subuid entry. Podman appears to hang after the fork while it is attempting to set up the user namespace. I've built my own binary with some additional logging and it seems to be hanging during the `setresuid` call. I'm not sure if it's waiting on a lock or something, but can reproduce easily if we need strace output.

For what it's worth, I'm able to run rootless runc containers on the same system, so I think the prerequisite kernel-level configuration is working.

**Steps to reproduce the issue:**

1. `export PODMAN_ALLOW_SINGLE_ID_MAPPING_IN_USERNS=1`

2. `podman info`

**Describe the results you received:**

No output, the process just hangs.

**Describe the results you expected:**

Success, or at the very least an error message if something failed.

**Output of `podman version`:**

```
Version:       0.9.3
Go Version:    go1.11
OS/Arch:       linux/amd64
```

**Output of `podman info`:**

```
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Physical RHEL7.5

**Comment by rhatdan** [comment]
@giuseppe this looks like it is up your alley.

**Comment by giuseppe** [comment]
`PODMAN_ALLOW_SINGLE_ID_MAPPING_IN_USERNS` is not really supported and should be used only for tests, for what are you using it?

**Comment by llchan** [comment]
I don't have root privileges so need to run rootless, and I also do not have permissions to modify /etc/subuid so had to set that env var. Adding each user to /etc/subuid also doesn't seem like a scalable solution on a multi-tenant host where usernames/uids are coming from NIS or whatever (sorry if terminology isnt quite right), so I think single-uid mode is something we should support.

As I mentioned above, my containers are working with a single uid via runc. I'm mostly hoping to use podman/buildah as higher-level tools to manage containers/images.

**Comment by giuseppe** [comment]
I still think the single mapping case should not be supported, as almost every command will fail if there are not enough mappings available and podman cannot pull/checkout any image with more than one ID.

Altough, the hang is a bug and it should never happen.

I've opened a PR to address it: https://github.com/containers/libpod/pull/1628

it is an "interesting" interaction of the Go runtime with libc versions that used to store the current uid/gid in a TLS variable.  Basically the Go runtime messes up some of the signals used by libc to notify all the threads that the uid/gid changed, so this never happens on setresuid/setresgid and the program hangs.  Since we have only one thread and we are going to re-exec immediately, we can skip the libc implementation completely and use directly the syscall

**Comment by llchan** [comment]
Thanks, appreciate the fix. Looks like it no longer hangs :+1:.

Is this the right place to discuss the single-mapping case in general? Maybe I'm not familiar enough with podman to understand, but why does it require more than one uid?

**Comment by rhatdan** [comment]
It is not podman that requires more then one UID but the images it pulls.  Many images require multiple UIDS, for example running the apache web service with uid 60 inside of your user namespace.  Mariadb runs as a separate UID from root as well.

Single UID systems will only work in containers where everything runs as CONTAINER ROOT. 

So the problem with supporting use cases where you only have a single UID, is that the containers will fail randomly on things like pulling an image that wants to set a UID on disk that is different then root.

**Comment by rhatdan** [comment]
@ebiederman any ideas on supporting network based /etc/subuid and /etc/subgid?

**Comment by llchan** [comment]
@rhatdan yeah if we can get a network-based subuid/subgid that would work, though I imagine that's going to require kernel-level changes so ideally we'd also have a shorter-term solution.

I understand the limitations of a single-uid mode, but given that only some subset of images require multi-uid support, it seems that single-uid is a reasonable use case to support. It can be gated behind an opt-in flag and accompanied by a use-at-your-own-risk warning message so no one is caught unaware.

Anyways this is sort of diverging from the original issue, which is already resolved. Should we create another issue to continue the discussion? I think there are a few bullet points to discuss (e.g. skip lchowns when when images are pulled in single-uid mode, network-based subuid/subgid, etc).

**Comment by mheon** [comment]
A separate issue seems like a good idea, sure
